tests/CMakeLists.txt

```diff
@@ -1,31 +1,54 @@
-  )
-endif()
+# Helper so we don't have to repeat ourselves so much
+function(add_cpp_test TEST_NAME TEST_FILE)
+  add_executable(${TEST_NAME} ${TEST_FILE})
+  add_test(${TEST_NAME} ${TEST_NAME})
+endfunction(add_cpp_test)
 
-add_cpp_test(basictests)
-add_cpp_test(errortests)
+# Sets a target to only build when you run the test, and expect failure
+function(add_compile_test TEST_NAME TEST_FILE EXPECT_SUCCESS)
+  add_executable(${TEST_NAME} ${TEST_FILE})
+  set_target_properties(${TEST_NAME} PROPERTIES
+                        EXCLUDE_FROM_ALL TRUE
+                        EXCLUDE_FROM_DEFAULT_BUILD TRUE)
+  add_test(NAME ${TEST_NAME}
+           COMMAND ${CMAKE_COMMAND} --build . --target ${TEST_NAME} --config $<CONFIGURATION>
+           WORKING_DIRECTORY ${CMAKE_BINARY_DIR})
+  if(NOT ${EXPECT_SUCCESS})
+    set_tests_properties(${TEST_NAME} PROPERTIES WILL_FAIL TRUE)
+  endif()
+endfunction(add_compile_test)
 
-## Next bit should not be needed!
-#if(CMAKE_INTERPROCEDURAL_OPTIMIZATION)
-# next line is a workaround for an odr-violation in basictests regarding the globals 0x432a40 and 0x52045c under clang
-#set_tests_properties(basictests PROPERTIES
-#    ENVIRONMENT ASAN_OPTIONS="detect_odr_violation=0")
-#endif()
+#
+# These test explicitly do #include "simdjson.cpp"
+#
+if (NOT MSVC) # Can't get simdjson-source to compile on Windows for some reason.
+  add_cpp_test(numberparsingcheck numberparsingcheck.cpp)
+  target_link_libraries(numberparsingcheck simdjson-source simdjson-windows-headers)
+  add_cpp_test(stringparsingcheck stringparsingcheck.cpp)
+  target_link_libraries(stringparsingcheck simdjson-source simdjson-windows-headers)
+endif()
 
-add_cpp_test(jsoncheck)
-add_cpp_test(parse_many_test)
-add_cpp_test(pointercheck)
-add_cpp_test(integer_tests)
+#
+# All remaining tests link with simdjson proper
+#
+link_libraries(simdjson)
 
-target_compile_definitions(basictests PRIVATE JSON_TEST_PATH="${PROJECT_SOURCE_DIR}/jsonexamples/twitter.json" NDJSON_TEST_PATH="${PROJECT_SOURCE_DIR}/jsonexamples/amazon_cellphones.ndjson")
-target_compile_definitions(errortests PRIVATE JSON_TEST_PATH="${PROJECT_SOURCE_DIR}/jsonexamples/twitter.json")
+# add_executable(allparserscheckfile allparserscheckfile.cpp)
+add_cpp_test(basictests basictests.cpp)
+add_cpp_test(errortests errortests.cpp)
+add_cpp_test(integer_tests integer_tests.cpp)
+add_cpp_test(jsoncheck jsoncheck.cpp)
+target_link_libraries(jsoncheck simdjson-windows-headers)
+add_cpp_test(parse_many_test parse_many_test.cpp)
+target_link_libraries(parse_many_test simdjson-windows-headers)
+add_cpp_test(pointercheck pointercheck.cpp)
 
-## This causes problems
-# add_executable(singleheader ./singleheadertest.cpp ${PROJECT_SOURCE_DIR}/singleheader/simdjson.cpp)
-# target_link_libraries(singleheader ${SIMDJSON_LIB_NAME})
-# add_test(singleheader singleheader)
+# Compile-only tests
+add_compile_test(readme_examples readme_examples.cpp TRUE)
```

Comment:
So instead of excluding the readme tests in VS 2017, and having someone who uses Visual Studio 2017 (the actual IDE, not in CI) swear at the failures... why don't we disable them directly in CMake...

```
if( "${CMAKE_GENERATOR}" MATCHES "Visual Studio" )
	if( MSVC_VERSION GREATER_EQUAL 1910 )
               # we have VS 2017 or better
               # do not add the unsupported tests
	elseif( MSVC_VERSION EQUAL 1920 )
		# we have VS 2019 or better
                # add the tests here
	else()
		message(FATAL_ERROR "Unsupported Visual Studio version")
	endif()
endif()
```

This pushes the complexity into the CMake scripts... which is not great, but it will be a quality-of-life improvements for users of Visual Studio 2017.

Reply:
Sure!